ramips: add support for Fon FON2601

drvlabo opened the pull request:
FON2601 is a wireless router.

Specification:
- SoC: Mediatek MT7620A (580MHz)
- RAM: 128 MiB
- ROM: 16 MiB SPI Flash
- Wireless:
   for 11b/g/n (upto 300 Mbps):  MT7620A built-in WMAC
   for 11a/n/ac (upto 867 Mbps): MT7662E
- Ethernet LAN: 1 port, upto 100 Mbps
- Ethernet WAN: 1 port, upto 1000 Mbps
- USB: 1 port (USB 2.0 host)
- LEDs: 4 (all can be controlled by SoC's GPIO)
- buttons: 1 (Displayed as "WPS" on enclosure)
- serial port: 57600n8
         pins: Vcc(3.3V), Rx, Tx, GND
        (left to right, viewed from outside of board)
    
Installation (only available via UART):
  1. download sysupgrade binary image by wget command
  2. write sysupgrade binary image to Flash
     command is:
       mtd write sysupgrade.bin firmware
  3. reboot
    
Important Notice:
  Only one button is displayed as "WPS" on enclosure.
  However, it is configured as "reset" (factory resetting feature).

Signed-off-by: NOGUCHI Hiroshi <drvlabo@gmail.com>

chunkeey commented:
It looks like the commits in https://github.com/openwrt/openwrt/pull/1834, https://github.com/openwrt/openwrt/pull/1833 and maybe https://github.com/openwrt/openwrt/pull/1835 belong to this PR. If this is the case, can you please integrate the patches into this PR and close the other ones? Thanks.

drvlabo commented:
@chunkeey Thanks for comments.
Merged modifications in #1833 and #1834. There are concerned to this PR.
#1835 does not belong to this PR.

chunkeey commented:
> #1835 does not belong to this PR.

Ok, so does your this PR depend on having #1835 then?

chunkeey commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -391,7 +391,8 @@ ramips_setup_interfaces()
 	wt1520-4M|\
-	wt1520-8M)
+	wt1520-8M|\
+	fon,fon2601)
```
Please change this. The entries should be in lexicographical order.

drvlabo replied:
changed at def4fb1

chunkeey commented on target/linux/ramips/dts/FON2601.dts:
```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-2.0
```
From what I know, upstream prefers ```GPL-2.0 OR MIT```, ```GPL-2.0-or-later OR MIT``` .

That said, it's totally fine if you keep it as is.

drvlabo replied:
Please let me keep it.

chunkeey replied:
Ok. GPL-2.0 it is.

chunkeey commented on target/linux/ramips/dts/FON2601.dts:
```diff
@@ -0,0 +1,167 @@
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset_wps {
```
Just a question: Is this is a multi-function key? Or really just reset?

drvlabo replied:
changed to single reset function button at cde53a9 .
This button is displayed as WPS on the enclosure.
However, factory resetting feature is prior to WPS and WPS feature cannot be configured easily by Luci. Therefore, I think that the button is used to dedicated "reset".

chunkeey replied:
> This button is displayed as WPS on the enclosure.

Oh, in that case please make this a WPS button. Otherwise this could confuse (end-)users that
grew accustom to the stock-firmware buttons usage. 

The ```hostapd-utils``` package contains the wps button support script. It should be enough to install the "full" wpad (as the -basic or -mini variants do not have WPS enabled) and said package in order to use the push-button feature. (though, the wifi config has to be edited to enable WPS too).

Alternatively,  if you are hell bent on keeping this as the factory reset. Then you should at least document that the buttons function was changed. (I do understand that a factory reset button is important!).

chunkeey commented on target/linux/ramips/dts/FON2601.dts:
```diff
@@ -0,0 +1,167 @@
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
Device node names should be as generic as possible. Please change this to "flash@0".

drvlabo replied:
changed at def4fb1

chunkeey commented on target/linux/ramips/dts/FON2601.dts:
```diff
@@ -0,0 +1,167 @@
+				read-only;
+			};
+
+			partition@1 {
```
The unit-address (the number after the "@") is should match the specified address-cell in the nodes' ```reg``` property . In this case it should be "partition@30000". 

(this needs also to be changed for the following partitions)

drvlabo replied:
changed at def4fb1

chunkeey commented on target/linux/ramips/dts/FON2601.dts:
```diff
@@ -0,0 +1,167 @@
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
```
I think it should be possible to directly reference the existing "state_default" label here.

&state_default {
     ...
};

(this will save one tab level because the &pinctrl { ... } gets removed ).

drvlabo replied:
changed at def4fb1

chunkeey replied:
Looks to me that ```&pinctrl {``` is still around? You should be able to drop it.

chunkeey commented on target/linux/ramips/dts/FON2601.dts:
```diff
@@ -0,0 +1,167 @@
+	status = "okay";
+};
+&pcie0 {
+	mt76@0,0 {
```
I think "wifi@0,0" would be a better fit.

This is listed in the example of the [mt76.txt binding](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/wireless/mediatek%2Cmt76.txt)

drvlabo replied:
changed at def4fb1

chunkeey replied:
Looks like the wifi@0,0 compatible is still missing? Please take a look at the MIR3G.dts for reference. (It has a 7662 on pcie1)

chunkeey commented on target/linux/ramips/image/Makefile:
```diff
@@ -130,6 +130,11 @@ define Build/zyimage
 endef
 
+define Build/fonfxcimage
+	$(STAGING_DIR_HOST)/bin/uimage_padhdr -i $@ -o $@.new
```
"$(STAGING_DIR_HOST)/bin/" might not be necessary anymore.

drvlabo replied:
changed at def4fb1

chunkeey commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -290,6 +290,19 @@ define Device/ex3700-ex3800
 endef
 TARGET_DEVICES += ex3700-ex3800
 
+define Device/fon2601
```
Please try: ```define Device/fon_fon2601```

This should also make the SUPPORTED_DEVICES entry unnecessary.

drvlabo replied:
changed at def4fb1

chunkeey commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -290,6 +290,19 @@ define Device/ex3700-ex3800
+			fonfxcimage |\
+			pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
+endef
+TARGET_DEVICES += fon2601
```
```TARGET_DEVICES += fon_fon2601```

drvlabo replied:
changed at def4fb1

drvlabo commented:
> > #1835 does not belong to this PR.
> 
> Ok, so does your this PR depend on having #1835 then?
No. #1835 is commonly concerned to all MT7620A targets which uses both embedded switch and external PHYs.
If #1835 will be rejected, we cannot use WAN port on this device. But, there is no more spoils.

chunkeey commented:
> If #1835 will be rejected, we cannot use WAN port on this device. But, there are no other spoils.

Ok, that's what I wanted to know. 

Can you please squash your (changes) patches into your PR patch? If you need help please look at the submitting-patches topic in the wiki:

https://openwrt.org/submitting-patches?s[]=squash#squashing_commits

drvlabo commented:
Patch squashing has done.
About a button, keeping factory resetting feature and added "Important Notice" into the last of commit comment.
And changed to fix for the other suggestions.

chunkeey commented on tools/firmware-utils/Makefile:
```diff
@@ -85,6 +85,7 @@ define Host/Compile
 	$(call cc,mkdlinkfw mkdlinkfw-lib, -lz -Wall --std=c99)
 	$(call cc,dns313-header, -Wall)
 	$(call cc,mksercommfw, -Wall)
+	$(call cc,uimage_padhdr, -lz)
```
No -Wall ?

drvlabo replied:
added -Wall flag.

ynezz replied:
Hm, where? :-)

ynezz replied:
```cc
gcc-9 -O2 -Wall -Werror -Wextra -Wno-unused-parameter ... uimage_padhdr.c  -lz

uimage_padhdr.c: In function ‘main’:
uimage_padhdr.c:138:10: error: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘long int’} and ‘long unsigned int’ [-Werror=sign-compare]
  138 |  if (rsz != (statbuf.st_size - sizeof(*imgh))) {
      |          ^~
uimage_padhdr.c:69:12: error: variable ‘crc_orig’ set but not used [-Werror=unused-but-set-variable]
   69 |  u_int32_t crc_orig;
      |            ^~~~~~~~
cc1: all warnings being treated as errors
```
```

chunkeey commented on target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_uimage.c:
```diff
@@ -376,6 +387,65 @@ static struct mtd_part_parser uimage_edimax_parser = {
+	crc_orig = be32_to_cpu(header->ih_hcrc);
+
+	header->ih_hcrc = 0;
+	crc_calc = uimage_crc32(0, buf,	sizeof(*header) + FONFXC_PAD_LEN);
```
I don't think you need to calculate and verify the crc at this point anymore. Like the other generic uImage parser and the Edimax you can just call uimage_verify_default() and leave it at that.

Reason is that the u-boot must have checked the correctness already otherwise it wouldn't have booted the kernel.

drvlabo replied:
If omitting header CRC check, we can not distinguish generic uimage header and fonfxc uimage header.
If actual length of ih_name string is less than 32 bytes, their only differences are the existence of zero padding (offset: +64-+95) and header CRC value.
The other header fields are all same means and same values.

chunkeey replied:
> If omitting header CRC check, we can not distinguish generic uimage header and fonfxc uimage header.

Then something (else) is wrong :confused: . The special compatible "fonfxc,uimage" is there to tell the mtdsplit which method to use. https://lists.openwrt.org/pipermail/openwrt-devel/2018-August/013558.html

so, looks like more work is required here. Question is: how to continue?

drvlabo replied:
May we assume that CONFIG_OF is enabled, in short, device tree bindings are activated ?
If so, only the parser for "fonfxc,uimage" can be directly called. In that case, CRC check is unnecessary, as you mentioned.

drvlabo replied:
Removed CRC check.

drvlabo commented:
@ynezz @chunkeey 
I have modified to fix all suggestions / pointings.
Check them please.

adrianschmutzler commented:
There has been a reorganization in ramips target today. Please:
- update your board.d subfiles since they have been re-sorted
- update your Makefile (mt****/rt****.mk) changes so sorting of devices is correct
- rename your DTS/DTSI files following the scheme soc_vendor_device.dts(i), e.g. mt7620a_tplink_archer-c50-v1.dts (and update includes there if necessary)
- remove `DTS := ...` from Makefile
- add `MTK_SOC` to Makefile (needs to match "soc" in DTS file name), e.g. `MTK_SOC := mt7620a`

Thank you!

Despite that, due to another recent patch you will have to split DEVICE_TITLE into DEVICE_VENDOR and DEVICE_MODEL.

Despite not having read the whole history, I would think about dividing your patches into logical units, so preparation of dependencies is separated from the actual device support patch.

drvlabo commented:
@adrianschmutzler 
Thanks for guide to fix. I did them.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -362,6 +362,20 @@ define Device/elecom_wrh-300cr
 
+define Device/fon_fon2601
+  MTK_SOC := mt7620a
+  IMAGE_SIZE := 16318464
```
Please use `15936k` here for consistency.

drvlabo replied:
modified.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -362,6 +362,20 @@ define Device/elecom_wrh-300cr
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs |\
+			fonfxcimage |\
+			pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
+  SUPPORTED_DEVICES += fon2601
```
If you add a new device, you do not need to add SUPPORTED_DEVICES, as the default will be the DTS compatible string.
Just remove this line.

drvlabo replied:
Removed. done.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_fon_fon2601.dts:
```diff
@@ -0,0 +1,167 @@
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
```
I think interrupt-driven "gpio-keys" is also available on mt7620a.
Try removing the poll-interval and using "gpio-keys" for compatible.

drvlabo replied:
Modified as that you mentioned.
Checked on my own target that interrupt-mode key is available.

drvlabo committed:
ramips: add support for Fon FON2601

FON2601 is a wireless router.

Specification:
- SoC: Mediatek MT7620A (580MHz)
- RAM: 128 MiB
- ROM: 16 MiB SPI Flash
- Wireless:
   for 11b/g/n (upto 300 Mbps):  MT7620A built-in WMAC
   for 11a/n/ac (upto 867 Mbps): MT7662E
- Ethernet LAN: 1 port, upto 100 Mbps
- Ethernet WAN: 1 port, upto 1000 Mbps
- USB: 1 port (USB 2.0 host)
- LEDs: 4 (all can be controlled by SoC's GPIO)
- buttons: 1 (Displayed as "WPS" on enclosure)
- serial port: 57600n8
 pins: Vcc(3.3V), Rx, Tx, GND
(left to right, viewed from outside of board)

Installation (only available via UART):
  1. download sysupgrade binary image by wget command
  2. write sysupgrade binary image to Flash
     command is:
       mtd write sysupgrade.bin firmware
  3. reboot

Important Notice:
  Only one button is displayed as "WPS" on enclosure.
  However, it is configured as "reset" (factory resetting feature).

Signed-off-by: NOGUCHI Hiroshi <drvlabo@gmail.com>

drvlabo commented:
Fixed them.

2019年7月25日(木) 5:34 Petr Štetiar <notifications@github.com>:

> *@ynezz* commented on this pull request.
> ------------------------------
>
> In tools/firmware-utils/Makefile
> <https://github.com/openwrt/openwrt/pull/1832#discussion_r307008446>:
>
> > @@ -85,6 +85,7 @@ define Host/Compile
>
>  	$(call cc,mkdlinkfw mkdlinkfw-lib, -lz -Wall --std=c99)
>
>  	$(call cc,dns313-header, -Wall)
>
>  	$(call cc,mksercommfw, -Wall)
>
> +	$(call cc,uimage_padhdr, -lz)
>
>
> gcc-9 -O2 -Wall -Werror -Wextra -Wno-unused-parameter ... uimage_padhdr.c  -lz
>
>
>
> uimage_padhdr.c: In function ‘main’:
>
> uimage_padhdr.c:138:10: error: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘long int’} and ‘long unsigned int’ [-Werror=sign-compare]
>
>   138 |  if (rsz != (statbuf.st_size - sizeof(*imgh))) {
>
>       |          ^~
>
> uimage_padhdr.c:69:12: error: variable ‘crc_orig’ set but not used [-Werror=unused-but-set-variable]
>
>    69 |  u_int32_t crc_orig;
>
>       |            ^~~~~~~~
>
> cc1: all warnings being treated as errors
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/1832?email_source=notifications&email_token=AIV46TZFSLX66YNMCWAD5A3QBC4FPA5CNFSM4GYCSQXKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7PPKGY#discussion_r307008446>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIV46T3IWW57RZVIHNN3ZELQBC4FPANCNFSM4GYCSQXA>
> .
>

drvlabo commented:
The latest commit is last commit for this PR.
If more modifications are required, I'll discard this PR.

I am tired of being forced to make corrections many times by being left uncommitted for a long time.

ynezz commented:
> by being left uncommitted for a long time.

Indeed, that's why I assigned this to myself and marked it `ready for merge` already, just need to find time to review/build/run test it on some mt7620.

> I am tired of being forced to make corrections many times 

I hear you, but if you're not going to fix them, then [someone else would need to do it anyway](https://git.openwrt.org/?p=openwrt/staging/ldir.git;a=shortlog;h=refs/heads/tools). Fixes in ` tools/firmware-utils` are quite problematic, because one usually has to deeply understand the image format and/or have device in order to run test it.

ynezz commented:
Merged, thanks!

ynezz commented:
@drvlabo FYI there's a [new bug report](https://bugs.openwrt.org/index.php?do=details&task_id=2413), that this PR has introduced regression on other device and we should try to fix it.

drvlabo committed:
Merge branch 'master' into fon2601_support

drvlabo commented:
If I'll fix that problem, should I create new another PR ?

drvlabo commented:
> FYI there's a new bug report,
When using mtdsplit w/o OF/device-tree, all mtdsplit parsers will be called from mtdparts.c:split_firmware().
Then, fonfxc parser (added by this PR) will be called at first in the parsers registered in mtdsplit_uimage.c .

fonfxc parser check top part of partition same as generic_uimage parser, so it returns SUCCESS and partition will be splitted with wrong offset position.

ynezz commented:
> If I'll fix that problem, should I create new another PR ?

I really don't care whether it's just another fix commit pushed to this PR or a new PR with fix commit, for me it's important the fix itself, thanks!

ynezz commented:
@drvlabo I've just attempted to fix one security issue found by Coverity Scan, could you please check it https://github.com/ynezz/openwrt/commit/29be761a4e172c904fe801947a0106b793ad8577 as well?

drvlabo commented:
@ynezz I checked your commit. I think it's good.
Thanks so much.

Secure coding is hard ...

ynezz commented:
Thanks, closing as it should be fixed via https://github.com/openwrt/openwrt/commit/e92a14709d37f64dcba8c81fa51e61e5f10f439a

adrianschmutzler commented:
@drvlabo Hi, would you possibly test whether my patch here produces working images for your device:
https://patchwork.ozlabs.org/project/openwrt/patch/20200710121623.1923-1-freifunk@adrianschmutzler.de/
